### ISSUE TITLE
Recompilation issues if highest_parent is not a model

### ISSUE THREAD

**Issue by javdrher** [GPflow/GPflow#442]
When adding models as children in other parentables, some issues arise with regards to recompling the likelihood graph. A small example illustrating this issue:
```python
import GPflow
import numpy as np

x = np.linspace(0, 2*np.pi, 10)[:,None]
y = np.sin(x)

m1 = GPflow.gpr.GPR(x, y, GPflow.kernels.RBF(1), GPflow.mean_functions.Zero())
m1.optimize()
m2 = GPflow.gpr.GPR(x.copy(), y.copy(), GPflow.kernels.RBF(1), GPflow.mean_functions.Zero())
m2.optimize()
p = GPflow.param.ParamList([m1, m2])
m1.likelihood.variance.fixed = True
m1.optimize()
```
The error is caused by line 692 in param.py, which sets `_needs_recompile` on the `ParamList` which is not picked up by optimize. I think this is unexpected behavior as strictly, only `Model` instances should have this attribute, as Parameterized objects don't have a _compile it shouldn't have the flag. In other words, only Model objects are sufficiently specialized to handle this flag. 
Usually, highest_parent is a model so the problem does not appear. In GPflowOpt however, the model sits in another parentable object.

Some potential fixes which could be developed: 
- In `optimize_np`and `optimize_tf` of `Model`, check for _needs_recompile of the highest_parent. In the example above however this would trigger recompilation on both m1 and m2, although only m1 needs to be.
- At line 692 in param.py, find the highest parent which is a `Model` (or has a _needs_recompile attribute which avoids a circular import of model.py in param.py). This would however invalidate some tests: more specifically testRecompile of ParamTestsDeeper, ParamTestsScalar and ParamTestsWider (in these tests no models are involved, yet the occurence of a _needs_recompile flag is tested after a change to the free state.

**Comment by markvdw** [comment]
This seems like a change in some core code, so I'd like to understand the problem a bit deeper before merging #456. Can you elaborate on your exact use case? So you say that that in GPflowOpt you have models in other parentable objects. Why is this, and do you want them to behave as separate objects or do you want their computational graph to interact with its parents?

If the answer that you want them to behave as separate objects, and you do not want the graphs to interact, I feel like there must be a better solution to this in `GPflowOpt`.

**Comment by javdrher** [comment]
The [Acquisition objects](https://github.com/GPflow/GPflowOpt/blob/master/GPflowOpt/acquisition/acquisition.py) in `GPflowOpt` act as transformations of the predictions. I therefor derived them from parameterized. They contain a list of models, and two AutoFlow functions to compute the acquisition functions (with and without gradients). Therefor, the computation of the acquisition function happens in TensorFlow, typically relying on `build_predict`, e.g. in the [EI class](https://github.com/GPflow/GPflowOpt/blob/master/GPflowOpt/acquisition/ei.py). 

Now if you go and use for instance, a VGP (like in GPflow/GPflowOpt#37) setting X and Y triggers a recompilation on the model. Unfortunately, this flag is set in the Acquisition object and so a call to `optimize`  doesn't recompile, causing a crash.

So i'd say the graphs interact

**Comment by markvdw** [comment]
Ok, so the way I see it, you never actually use the results of the compile function of the "child" models? You only use the child models to construct parts of the graph which are then used downstream in the `Acquisition` object?

**Comment by markvdw** [comment]
@javdrher and I discussed this offline and possibly found a workaround for now. @javdrher, if you feel that this has been resolved for now, could you close the issue?

Also, if this is still an issue in the future, I think all of us need a better understanding of exactly the use-case that this is necessary for. From our discussion, I gather that the main issue is that you want to use models to create sub-graphs for the class you're interested in, rather than be the "root node" which optimises everything. Currently only `Parameterized` classes allow this I think (like `Kern` and `Likelihood` objects). We're definitely thinking about implementing something like this in time, but a clear overview of exactly what we allow and the semantics of the classes is something that still needs work.

**Comment by javdrher** [comment]
@markvdw I tested what we discussed, but ran into a couple of issues. In addition to tf_mode, also most methods like get_free_state and make_tf_array were broken. I solved those by overriding sorted_params. 

Unfortunately, the issue I currently can't get around is _kill_autoflow, which doesn't start at the root anymore because the models no longer hold a parent reference. As a result, the AF storages of the evaluate functions in Acquisition aren't cleared.

**Comment by javdrher** [comment]
I managed to create a working version using the current GPflow master. However it requires a fix I classify as hack. Because models are wrapped in GPflowOpt I can intercept the highest_parent call in those wrappers and insert an object which is able to set the recompilation on the right object as well as trigger the removal of AF storages, [code here](https://github.com/GPflow/GPflowOpt/blob/recompilation_fix/GPflowOpt/models.py). 

It's a workaround I can temporarily live with as it allows us to proceed and at the same time think about this use case carefully, though I hope you'll understand I'd rather not keep this permanently.

As use-case definition, think of it as model integration at the tensorflow level, instead of numpy. We made the Acquisition part of the parentable hierarchy as it maps the predictions of one or more models in a specific way, using tensorflow. It's a way of constructing larger graphs, where under circumstances models are still being optimized. This works well, except for the fact that we violate an assumption in the recompilation scheme, being that the highest parentable is able to recompile. This assumption is implicit in GPflow (in a stronger typed language the highest parent would need to be cast to a model class).
